### PR TITLE
Updated OneDrive 2.1, 2.2 and 2.3 policy

### LINKED ISSUE #31: Agency 2 Pilot: OneDrive; 2.2 and 2.3 error if anyone links are disabled (Policy 2.1)
OneDrive for Business policy 2.1 is "Anyone Links SHOULD Be Turned Off". If this is turned off the current rego will mark 2.2 and 2.3 as "FAIL" (Expiration Date SHOULD Be Set for Anyone Links & Link Permissions SHOULD Be Set to Enabled Anyone Links to View) since the admin cannot set an expiration date or permissions to "view".

Suggest adding to the rego a "PASS" scenario if anyone links are disable 2.2 and 2.3 will also show up as "PASS".

### PR DESCRIPTION
**Pull request by Dylan-MITRE**
Modified OneDrive 2.1, 2.2 and 2.3 policy

## 🗣 Description ##

Modified OneDrive 2.2 and 2.3 policy, so that
1. When 2.1 pass, 2.2 and 2.3 automatically pass and indicated "Requirement met: AnyOne Link is disabled"
2. fixed the checks for policy 2.2 
3. 2.1 now pass for more restricted sharing settings

## 💭 Motivation and context ##

closes #31 #127 #154 

## 🧪 Testing ##

1. Open the Sharepoint Admin Center
2. Navigate to Policies > Sharing
3. Set the External Sharing slider to New and existing guests. (run test- 2.1, 2.2, 2.3 pass)
5. Set the External Sharing slider to Anyone. 
6. Scroll to the section "Choose expiration and permissions options for Anyone links." 
To test 2.2 
7. Uncheck "These links must expire within this many days" (run test -2.1 fail, 2.2 fails)
8. Uncheck "These links must expire within this many days" and input number thats not 30 (run test -2.1 fail, 2.2 fails)
9. Uncheck "These links must expire within this many days" and input number 30 (run test -2.1 fail, 2.2 pass)

To test 2.3 (tested in #112)
Mainly just to make sure when 2.1 pass, 2.3 pass

PS: Don't forget to revert the changes after testing

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

**Review by tkol2022 — CHANGES_REQUESTED** [review]
@Dylan-MITRE Please remove references to Get-SPOSite in the OneDrive provider because it is not being used by the OneDrive Rego.

**Comment by tkol2022** [comment]
@Dylan-MITRE In addition to the suggested revisions that @schrolla suggested, please make the following adjustments to the code:

- Change OneDriveLoopSharingCapability to SharingCapability. Let's talk verbally about why we are making this change because it is easier to explain that way.

- Remove DefaultSharingLinkType from the logic because that setting is irrelevant to this policy.


- OneDriveRequestFilesLinkExpirationInDays is not the correct setting to check for 2.2 policy 2. The correct setting is RequireAnonymousLinksExpireInDays (see the dependent GitHub issue).

**Comment by tkol2022** [comment]
@schrolla @Dylan-MITRE I believe this pull request addresses #127 which was opened a few days ago. Please double check the implementation for setting 2.1 in the repo associated with this PR and link that other issue if I'm correct. I see the code below which appears to be checking for any sharing slider value other than Anyone. 

```
#
# Baseline 2.1: Policy 1
#--
AnyoneLinksPolicy[Policy]{
    Policy := input.SPO_tenant_info[_]
    Policy.OneDriveLoopSharingCapability != 2
}

```

**Comment by Dylan-MITRE** [comment]
> #127

good catch i link #127 in the pull request and changed the PR title

File: Rego/OneDriveConfig.rego [hunk 1]
```diff
@@ -40,46 +40,80 @@ tests[{
...
+    "ActualValue" : [Policy.OneDriveLoopSharingCapability, Policy.RequireAnonymousLinksExpireInDays],
+    "ReportDetails" : ReportDetails2_2_1(Policy),
     "RequirementMet" : Status
 }] {
-    Policies := RequiredExpirationDatePolicy
-    Status := count(Policies) == 1
+    Policy := input.SPO_tenant_info[_]
+    Conditions1 := [Policy.OneDriveLoopSharingCapability !=2]
+    Case1 := count([Condition | Condition = Conditions1[_]; Condition == false]) == 0
+	Conditions2 := [Policy.OneDriveLoopSharingCapability == 2, Policy.RequireAnonymousLinksExpireInDays != -1]
```
Comment (Sloane4) [inline Rego/OneDriveConfig.rego:72]: There is extra spacing on this line

File: Rego/OneDriveConfig.rego [hunk 2]
```diff
@@ -40,46 +40,80 @@ tests[{
...
+    "ActualValue" : [Policy.OneDriveLoopSharingCapability, Policy.OneDriveRequestFilesLinkExpirationInDays],
+    "ReportDetails" : ReportDetails2_2_2(Policy),
     "RequirementMet" : Status
 }] {
-    Policies := ExpirationDatePolicy
-    Status := count(Policies) == 1
+    Policy := input.SPO_tenant_info[_]
+    Conditions1 := [Policy.OneDriveLoopSharingCapability !=2]
+    Case1 := count([Condition | Condition = Conditions1[_]; Condition == false]) == 0
+	Conditions2 := [Policy.OneDriveLoopSharingCapability == 2, Policy.OneDriveRequestFilesLinkExpirationInDays == 30]
```
Comment (Sloane4) [inline Rego/OneDriveConfig.rego:112]: There is extra spacing on this line

File: Rego/OneDriveConfig.rego [hunk 3]
```diff
@@ -90,50 +124,54 @@ tests[{
...
     "Commandlet" : ["Get-SPOTenant"],
-    "ActualValue" : [Policy.DefaultSharingLinkType, Policy.FileAnonymousLinkType, Policy.FolderAnonymousLinkType],
+    "ActualValue" : [Policy.OneDriveLoopSharingCapability, Policy.FileAnonymousLinkType, Policy.FolderAnonymousLinkType],
     "ReportDetails" : ReportDetails2_3(Policy),
     "RequirementMet" : Status
 }] {
     Policy := input.SPO_tenant_info[_]
-	Conditions := [Policy.DefaultSharingLinkType != 3, Policy.FileAnonymousLinkType == 1, Policy.FolderAnonymousLinkType == 1]
-    Status := count([Condition | Condition = Conditions[_]; Condition == false]) == 0
+        Conditions1 := [Policy.OneDriveLoopSharingCapability !=2]
```
Comment (Sloane4) [inline Rego/OneDriveConfig.rego:169]: There is extra spacing on this line to line 173

**Review by Sloane4 — APPROVED** [review]
Just some nit picky stuff

File: Rego/OneDriveConfig.rego (hunk 1 above)
Comment (Dylan-MITRE) [reply Rego/OneDriveConfig.rego]: thank you for pointing it out, changed

File: Rego/OneDriveConfig.rego (hunk 2 above)
Comment (Dylan-MITRE) [reply Rego/OneDriveConfig.rego]: thank you for pointing it out, changed

File: Rego/OneDriveConfig.rego (hunk 3 above)
Comment (Dylan-MITRE) [reply Rego/OneDriveConfig.rego]: thank you for pointing it out, changed

**Comment by tkol2022** [comment]
Tested the case where the OneDrive slider is set to any value more restrictive than Anyone and it shows a Pass as expected.

**Comment by tkol2022** [comment]
### Testing from the Sharepoint Admin center portal
Tested all possible permutations of the 2.2 and 2.3 settings below when Anyone links are enabled and disabled to ensure that the report outputs the expected result. Looks good when setting the configurations from the portal.

image

**Comment by tkol2022** [comment]
As you modify the Rego code (and unit tests) for the pending updates, change all instances of the OneDriveLoopSharingCapability field to OneDriveSharingCapability instead. Hopefully this can be a mass replace and should not affect the structure of any policies that you've coded. See new issue #154 for details.

**Comment by Dylan-MITRE** [comment]
Notes for myself: 

Items to check:
- [x] OneDriveLoopSharingCapability to OneDriveSharingCapability
- [ ] RequireAnonymousLinksExpireInDays
- [ ] OneDriveRequestFilesLinkExpirationInDays

**Comment by Dylan-MITRE** [comment]
@tkol2022 changed 2.2 based on your feedback on RequireAnonymousLinksExpireInDays and OneDriveRequestFilesLinkExpirationInDays, changed 2.1, 2.2, 2.3 based on your feedback of OneDriveSharingCapability.
can you double check if I missed anything

**Comment by tkol2022** [comment]
### This comment includes the details of the final round of testing after all code updates.
**Every test passed.** Nice work. Details of each case that I ran is included below for future reference.

### Test 1 - Make sure 2.1, 2.2, 2.3 pass when external sharing with Anyone is off
Set the value of Anyone links expiration date (2.2) to a non compliant value. Set the value of file permissions (2.3) to a non compliant value. Iteratively set the value of OneDriveSharingCapability to all values except Anyone. 

**New and existing guests** 
Set-SPOTenant -OneDriveSharingCapability ExternalUserSharingOnly -RequireAnonymousLinksExpireInDays 100 -FileAnonymousLinkType Edit -FolderAnonymousLinkType Edit
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that 2.1, 2.2, 2.3 pass

**Existing guests** 
Set-SPOTenant -OneDriveSharingCapability ExistingExternalUserSharingOnly -RequireAnonymousLinksExpireInDays 100 -FileAnonymousLinkType Edit -FolderAnonymousLinkType Edit
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that 2.1, 2.2, 2.3 pass

**Only people in your organization** 
Set-SPOTenant -OneDriveSharingCapability Disabled -RequireAnonymousLinksExpireInDays 100 -FileAnonymousLinkType Edit -FolderAnonymousLinkType Edit
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that 2.1, 2.2, 2.3 pass


### Test 2 (2.2) - When Anyone links are on, make sure there is an expiration of 30 days set

Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -RequireAnonymousLinksExpireInDays 30
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.2 results show Pass


Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -RequireAnonymousLinksExpireInDays 100
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.2 results show Fail

### Test 3 (2.3) - When Anyone links are on, make sure that the file and folder permissions are View

Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -FileAnonymousLinkType View -FolderAnonymousLinkType View
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.3 results show Pass


Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -FileAnonymousLinkType Edit -FolderAnonymousLinkType View
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.3 results show Fail


Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -FileAnonymousLinkType View -FolderAnonymousLinkType Edit
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.3 results show Fail


Set-SPOTenant -OneDriveSharingCapability ExternalUserAndGuestSharing -FileAnonymousLinkType Edit -FolderAnonymousLinkType Edit
Invoke-SCuBA -ProductNames onedrive -LogIn $false

Check to ensure that the 2.3 results show Fail

**Review by tkol2022 — APPROVED** [review]
Please resolve the conflicts and this PR will be ready to merge.

**Comment by Dylan-MITRE** [comment]
@tkol2022 conflicted is resolved by adding in the pnp changes

**Pull request merged**